File: python/src/agreements_api/get_agreement_pricing_type/get_agreement_pricing_type.py
"""
Purpose
Obtain the pricing type of the agreement (contract, FPS, metered, free etc.)
AG-16
"""

import json
import logging

import boto3
from botocore.exceptions import ClientError

# To search by offer id: OfferId; by product id: ResourceIdentifier; by product type: ResourceType
idType = "OfferId"

# replace id value as needed
idValue = "offer-1111111111111"

MAX_PAGE_RESULTS = 10

# catalog; switch to AWSMarketplace for release
AWSMPCATALOG = "AWSMarketplace"

# product types

SaaSProduct = "SaaSProduct"
AmiProduct = "AmiProduct"
MLProduct = "MachineLearningProduct"
ContainerProduct = "ContainerProduct"
DataProduct = "DataProduct"
ProServiceProduct = "ProfessionalServicesProduct"
AiqProduct = "AiqProduct"

# Define pricing types
CCP = "CCP"
Annual = "Annual"
Contract = "Contract"
SFT = "SaaS Freee Trial"
HMA = "Hourly and Monthly Agreements"
Hourly = "Hourly"
Monthly = "Monthly"
AFPS = "Annual FPS"
CFPS = "Contract FPS"
CCPFPS = "CCP with FPS"
BYOL = "BYOL"
Free = "Free"
FTH = "Free Trials and Hourly"

# Define Agreement Term Types
legal = ["LegalTerm"]
config = ["ConfigurableUpfrontPricingTerm"]
usage = ["UsageBasedPricingTerm"]
config_usage = ["ConfigurableUpfrontPricingTerm", "UsageBasedPricingTerm"]
freeTrial = ["FreeTrialPricingTerm"]
recur = ["RecurringPaymentTerm"]
usage_recur = ("UsageBasedPricingTerm", "RecurringPaymentTerm")
fixed_payment = ["FixedUpfrontPricingTerm", "PaymentScheduleTerm"]
fixed_payment_usage = [
    "FixedUpfrontPricingTerm",
    "PaymentScheduleTerm",
    "UsageBasedPricingTerm",
]
byol = ["ByolPricingTerm"]
freeTrial_usage = ("FreeTrialPricingTerm", "UsageBasedPricingTerm")
all_agreement_types_combination = (
    legal,
    config,
    usage,
    config_usage,
    freeTrial,
    recur,
    usage_recur,
    fixed_payment,
    fixed_payment_usage,
    byol,
    freeTrial_usage,
)


# get pricing type method given product type, agreement temr type and offer type if needed
def get_pricing_type(product_type, agreement_term_type, offer_type):
    pricing_types = {
        (SaaSProduct, frozenset(config_usage), frozenset("")): CCP,
        (DataProduct, frozenset(config_usage), frozenset("")): CCP,
        (ContainerProduct, frozenset(config), frozenset(config_usage)): Annual,
        (AmiProduct, frozenset(config), frozenset(config_usage)): Annual,
        (MLProduct, frozenset(config), frozenset(config_usage)): Annual,
        (ContainerProduct, frozenset(config), frozenset(config)): Contract,
        (AmiProduct, frozenset(config), frozenset(config)): Contract,
        (SaaSProduct, frozenset(config), frozenset("")): Contract,
        (DataProduct, frozenset(config), frozenset("")): Contract,
        (AiqProduct, frozenset(config), frozenset("")): Contract,
        (ProServiceProduct, frozenset(config), frozenset("")): Contract,
        (SaaSProduct, frozenset(freeTrial), frozenset("")): SFT,
        (AmiProduct, frozenset(usage_recur), frozenset("")): HMA,
        (SaaSProduct, frozenset(usage), frozenset("")): Hourly,
        (AmiProduct, frozenset(usage), frozenset("")): Hourly,
        (ContainerProduct, frozenset(usage), frozenset("")): Hourly,
        (MLProduct, frozenset(usage), frozenset("")): Hourly,
        (ContainerProduct, frozenset(recur), frozenset("")): Monthly,
        (AmiProduct, frozenset(recur), frozenset("")): Monthly,
        (
            ContainerProduct,
            frozenset(fixed_payment),
            frozenset(fixed_payment_usage),
        ): AFPS,
        (AmiProduct, frozenset(fixed_payment), frozenset(fixed_payment_usage)): AFPS,
        (MLProduct, frozenset(fixed_payment), frozenset("")): AFPS,
        (ContainerProduct, frozenset(fixed_payment), frozenset(fixed_payment)): CFPS,
        (AmiProduct, frozenset(fixed_payment), frozenset(fixed_payment)): CFPS,
        (SaaSProduct, frozenset(fixed_payment), frozenset("")): CFPS,
        (DataProduct, frozenset(fixed_payment), frozenset("")): CFPS,
        (AiqProduct, frozenset(fixed_payment), frozenset("")): CFPS,
        (ProServiceProduct, frozenset(fixed_payment), frozenset("")): CFPS,
        (SaaSProduct, frozenset(fixed_payment_usage), frozenset("")): CCPFPS,
        (DataProduct, frozenset(fixed_payment_usage), frozenset("")): CCPFPS,
        (AiqProduct, frozenset(fixed_payment_usage), frozenset("")): CCPFPS,
        (ProServiceProduct, frozenset(fixed_payment_usage), frozenset("")): CCPFPS,
        (AmiProduct, frozenset(byol), frozenset("")): BYOL,
        (SaaSProduct, frozenset(byol), frozenset("")): BYOL,
        (ProServiceProduct, frozenset(byol), frozenset("")): BYOL,
        (AiqProduct, frozenset(byol), frozenset("")): BYOL,
        (MLProduct, frozenset(byol), frozenset("")): BYOL,
        (ContainerProduct, frozenset(byol), frozenset("")): BYOL,
        (DataProduct, frozenset(byol), frozenset("")): BYOL,
        (ContainerProduct, frozenset(legal), frozenset("")): Free,
        (AmiProduct, frozenset(freeTrial_usage), frozenset("")): FTH,
        (ContainerProduct, frozenset(freeTrial_usage), frozenset("")): FTH,
        (MLProduct, frozenset(freeTrial_usage), frozenset("")): FTH,
    }

    key = (product_type, agreement_term_type, offer_type)
    if key in pricing_types:
        return pricing_types[key]
    else:
        return "Unknown"


# Example usage for testing purpose
"""
product_type = SaaSProduct
agreement_term_type = frozenset(config_usage)
offer_type = frozenset('')
pricing_type = get_pricing_type(product_type, agreement_term_type, offer_type)
print("pricing type = " + pricing_type)  # Output: CCP
"""


# check if offer term types are needed; if Y, needed
def get_offer_term_type(product_type, agreement_term_type):
    offer_term_types = {
        (ContainerProduct, frozenset(config)): "Y",
        (AmiProduct, frozenset(config)): "Y",
        (ContainerProduct, frozenset(fixed_payment)): "Y",
        (AmiProduct, frozenset(fixed_payment)): "Y",
        (AmiProduct, frozenset(fixed_payment), frozenset(fixed_payment)): "Y",
    }

    key = (product_type, agreement_term_type)
    if key in offer_term_types:
        return offer_term_types[key]
    else:
        return


logger = logging.getLogger(__name__)


def get_agreements(mp_client):
    AgreementSummaryList = []
    partyTypes = ["Proposer"]
    for value in partyTypes:
        try:
            agreement = mp_client.search_agreements(
                catalog=AWSMPCATALOG,
                maxResults=MAX_PAGE_RESULTS,
                filters=[
                    {"name": "PartyType", "values": [value]},
                    {"name": idType, "values": [idValue]},
                    {"name": "AgreementType", "values": ["PurchaseAgreement"]},
                ],
            )
        except ClientError as e:
            logger.error("Could not complete search_agreements request.")
            raise

        AgreementSummaryList.extend(agreement["agreementViewSummaries"])

        while "nextToken" in agreement and agreement["nextToken"] is not None:
            try:
                agreement = mp_client.search_agreements(
                    catalog=AWSMPCATALOG,
                    maxResults=MAX_PAGE_RESULTS,
                    nextToken=agreement["nextToken"],
                    filters=[
                        {"name": "PartyType", "values": [value]},
                        {"name": idType, "values": [idValue]},
                        {"name": "AgreementType", "values": ["PurchaseAgreement"]},
                    ],
                )
            except ClientError as e:
                logger.error("Could not complete search_agreements request.")
                raise

            AgreementSummaryList.extend(agreement["agreementViewSummaries"])

    return AgreementSummaryList


def usage_demo():
    logging.basicConfig(level=logging.INFO, format="%(levelname)s: %(message)s")

    print("-" * 88)
    print("Looking for an agreement in the AWS Marketplace Catalog.")
    print("-" * 88)

    mp_client = boto3.client("marketplace-agreement")

    # find all agreements matching the specified idType and idValue
    agreements = get_agreements(mp_client)

    for item in agreements:
        pricingType = ""
        agreement_id = item["agreementId"]

        # get term types inside offer
        offer_term_types = get_offer_term_types(item)

        # even though multiple product types are allowed for one agreement, only need the first one
        productType = item["resourceSummaries"][0]["resourceType"]

        # get agreement terms types
        agreementTerm = mp_client.get_agreement_terms(agreementId=agreement_id)

        agreementTermTypes = get_agreement_term_types(agreementTerm)

        # match with agreement term type group
        matchedTermType = getMatchedTermTypesCombination(agreementTermTypes)

        # check if offer term type is needed.
        offer_term_type_needed = get_offer_term_type(
            productType, frozenset(matchedTermType)
        )

        # get pricing type given product type, agreement term types and offer type if needed;
        # one excpetion is Container with Legal term. LegalTerm needs to be the only term present
        if offer_term_type_needed is not None:
            matchedOfferTermTypes = getMatchedTermTypesCombination(offer_term_types)
            print(f"matchedOfferTermType = {matchedOfferTermTypes}")
            pricingType = get_pricing_type(
                productType,
                frozenset(matchedTermType),
                frozenset(matchedOfferTermTypes),
            )
        elif set(matchedTermType) == set(legal):
            pricingType = Free
        else:
            pricingType = get_pricing_type(
                productType, frozenset(matchedTermType), frozenset("")
            )

        print(
            f"agreementId={agreement_id};productType={productType}; agreementTermTypes={agreementTermTypes}; matchedTermType={matchedTermType}; offerTermTypeNeeded={offer_term_type_needed}; offer_term_types={offer_term_types}"
        )
        print(f"pricing type={pricingType}")


def getMatchedTermTypesCombination(agreementTermTypes):
    matchedCombination = ()
    for element in all_agreement_types_combination:
        if check_elements(agreementTermTypes, element):
            matchedCombination = element
    return matchedCombination


def get_offer_term_types(item):
    offer_id = item["agreementTokenSummary"]["offerId"]
    mp_catalogAPI_client = boto3.client("marketplace-catalog")
    offer_document = get_entity_information(mp_catalogAPI_client, offer_id)
    offerDetail = offer_document["Details"]
    offerDetail_json_object = json.loads(offerDetail)
    offer_term_types = [term["Type"] for term in offerDetail_json_object["Terms"]]
    return offer_term_types


# make sure all elements in array2 exist in array1
def check_elements(array1, array2):
    for element in array2:
        if element not in array1:
            return False
    return True


def get_entity_information(mp_client, entity_id):
    """
    Returns information about a given entity
    Args: entity_id str: Entity to return
    Returns: dict: Dictionary of entity information
    """

    try:
        response = mp_client.describe_entity(
            Catalog="AWSMarketplace",
            EntityId=entity_id,
        )

        return response

    except ClientError as e:
        if e.response["Error"]["Code"] == "ResourceNotFoundException":
            logger.error("Entity with ID %s not found.", entity_id)
        else:
            logger.error("Unexpected error: %s", e)


def get_agreement_term_types(agreementTerm):
    types = []
    for term in agreementTerm["acceptedTerms"]:
        for value in term.values():
            if isinstance(value, dict) and "type" in value:
                types.append(value["type"])
    return types


if __name__ == "__main__":
    usage_demo()

File: java/src/main/java/com/amazon/samplelib/agreement/GetAgreementPricingType.java
//AG-18 Obtain the EULA I have entered into with my customer via the agreement
package com.amazon.samplelib.agreement;

import software.amazon.awssdk.services.marketplaceagreement.MarketplaceAgreementClient;
import software.amazon.awssdk.services.marketplaceagreement.model.AcceptedTerm;
import software.amazon.awssdk.services.marketplaceagreement.model.AgreementViewSummary;
import software.amazon.awssdk.services.marketplaceagreement.model.Filter;
import software.amazon.awssdk.services.marketplaceagreement.model.GetAgreementTermsRequest;
import software.amazon.awssdk.services.marketplaceagreement.model.GetAgreementTermsResponse;
import software.amazon.awssdk.services.marketplaceagreement.model.SearchAgreementsRequest;
import software.amazon.awssdk.services.marketplaceagreement.model.SearchAgreementsResponse;

import com.fasterxml.jackson.annotation.JsonAutoDetect.Visibility;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Objects;
import java.util.Set;

import org.apache.commons.lang3.tuple.Triple;

import software.amazon.awssdk.services.marketplacecatalog.MarketplaceCatalogClient;
import software.amazon.awssdk.services.marketplacecatalog.model.DescribeEntityRequest;
import software.amazon.awssdk.services.marketplacecatalog.model.DescribeEntityResponse;

import com.amazon.samplelib.utils.ReferenceCodesConstants;
import com.amazon.samplelib.utils.ReferenceCodesUtils;
import com.fasterxml.jackson.annotation.PropertyAccessor;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.databind.ObjectWriter;
import com.fasterxml.jackson.datatype.jsr310.JavaTimeModule;

public class GetAgreementPricingType {
	
	private static final String idType = "OfferId";
	
	private static final String idValue = ReferenceCodesConstants.OFFER_ID;
	
	private static MarketplaceAgreementClient mpasClient = ReferenceCodesUtils.getMPAgreementClient();

	private static MarketplaceCatalogClient awsMarketplaceCatalog = ReferenceCodesUtils.getMPCatalogClient();

	  // Product types
    private static final String SaaSProduct = "SaaSProduct";
    private static final String AmiProduct = "AmiProduct";
    private static final String MLProduct = "MachineLearningProduct";
    private static final String ContainerProduct = "ContainerProduct";
    private static final String DataProduct = "DataProduct";
    private static final String ProServiceProduct = "ProfessionalServicesProduct";
    private static final String AiqProduct = "AiqProduct";

    // Pricing types
    private static final String CCP = "CCP";
    private static final String Annual = "Annual";
    private static final String Contract = "Contract";
    private static final String SFT = "SaaS Free Trial";
    private static final String HMA = "Hourly and Monthly Agreements";
    private static final String Hourly = "Hourly";
    private static final String Monthly = "Monthly";
    private static final String AFPS = "Annual FPS";
    private static final String CFPS = "Contract FPS";
    private static final String CCPFPS = "CCP with FPS";
    private static final String BYOL = "BYOL";
    private static final String Free = "Free";
    private static final String FTH = "Free Trials and Hourly";    

    // Agreement term types
    private static final Set<String> legal = Set.of("LegalTerm");
    private static final Set<String> config = Set.of("ConfigurableUpfrontPricingTerm");
    private static final Set<String> usage = Set.of("UsageBasedPricingTerm");
    private static final Set<String> configUsage = Set.of("ConfigurableUpfrontPricingTerm","UsageBasedPricingTerm"); 
    private static final Set<String> freeTrial = Set.of("FreeTrialPricingTerm");
    private static final Set<String> recur = Set.of("RecurringPaymentTerm");
    private static final Set<String> usageRecur = Set.of("UsageBasedPricingTerm","RecurringPaymentTerm");
    private static final Set<String> fixedPayment = Set.of("FixedUpfrontPricingTerm", "PaymentScheduleTerm");
    private static final Set<String> fixedPaymentUsage = Set.of("FixedUpfrontPricingTerm", "PaymentScheduleTerm", "UsageBasedPricingTerm");
    private static final Set<String> byol = Set.of("ByolPricingTerm");
    private static final Set<String> freeTrialUsage = Set.of("FreeTrialPricingTerm", "UsageBasedPricingTerm");

    private static List<Set<String>> allAgreementTypesCombination = Arrays.asList(
    		legal,config, usage, configUsage, freeTrial, recur, usageRecur, fixedPayment, fixedPaymentUsage, byol, freeTrialUsage	
    );
    
    

    public static String getPricingType(String productType, Set<String> agreementTermType, Set<String> offerType) {
        Map<Triple<String, Set<String>, Set<String>>, String> pricingTypes = new HashMap<>();
        
        pricingTypes.put(Triple.of(SaaSProduct, configUsage, new HashSet<>()), CCP);
        pricingTypes.put(Triple.of(DataProduct, configUsage, new HashSet<>()), CCP);
        pricingTypes.put(Triple.of(ContainerProduct, config, configUsage), Annual);
        pricingTypes.put(Triple.of(AmiProduct, config, configUsage), Annual);
        pricingTypes.put(Triple.of(MLProduct, config, configUsage), Annual);
        pricingTypes.put(Triple.of(ContainerProduct, config, config), Contract);
        pricingTypes.put(Triple.of(AmiProduct, config, config), Contract);
        pricingTypes.put(Triple.of(SaaSProduct, config,  new HashSet<>()), Contract);
        pricingTypes.put(Triple.of(DataProduct, config,  new HashSet<>()), Contract);
        pricingTypes.put(Triple.of(AiqProduct, config,  new HashSet<>()), Contract);
        pricingTypes.put(Triple.of(ProServiceProduct, config,  new HashSet<>()), Contract);
        pricingTypes.put(Triple.of(SaaSProduct, freeTrial,  new HashSet<>()), SFT);
        pricingTypes.put(Triple.of(AmiProduct, usageRecur,  new HashSet<>()), HMA);
        pricingTypes.put(Triple.of(SaaSProduct, usage,  new HashSet<>()), Hourly);
        pricingTypes.put(Triple.of(AmiProduct, usage,  new HashSet<>()), Hourly);
        pricingTypes.put(Triple.of(MLProduct, usage,  new HashSet<>()), Hourly);
        pricingTypes.put(Triple.of(ContainerProduct, recur,  new HashSet<>()), Monthly);
        pricingTypes.put(Triple.of(AmiProduct, recur,  new HashSet<>()), Monthly);
        pricingTypes.put(Triple.of(ContainerProduct, fixedPayment,  fixedPaymentUsage), AFPS);
        pricingTypes.put(Triple.of(AmiProduct, fixedPayment, fixedPaymentUsage), AFPS);
        pricingTypes.put(Triple.of(MLProduct, fixedPayment, new HashSet<>()),AFPS);
        pricingTypes.put(Triple.of(ContainerProduct, fixedPayment, new HashSet<>()),CFPS);
        pricingTypes.put(Triple.of(AmiProduct, fixedPayment, fixedPayment),CFPS);
        pricingTypes.put(Triple.of(SaaSProduct, fixedPayment, new HashSet<>()),CFPS);
        pricingTypes.put(Triple.of(DataProduct, fixedPayment, new HashSet<>()),CFPS);
        pricingTypes.put(Triple.of(AiqProduct, fixedPayment, new HashSet<>()),CFPS);
        pricingTypes.put(Triple.of(ProServiceProduct, fixedPayment, new HashSet<>()),CFPS);
        pricingTypes.put(Triple.of(SaaSProduct, fixedPaymentUsage, new HashSet<>()),CCPFPS);
        pricingTypes.put(Triple.of(DataProduct, fixedPaymentUsage, new HashSet<>()),CCPFPS);
        pricingTypes.put(Triple.of(AiqProduct, fixedPaymentUsage, new HashSet<>()),CCPFPS);
        pricingTypes.put(Triple.of(ProServiceProduct, fixedPaymentUsage, new HashSet<>()),CCPFPS);
        pricingTypes.put(Triple.of(AmiProduct, byol, new HashSet<>()),BYOL);
        pricingTypes.put(Triple.of(SaaSProduct, byol, new HashSet<>()),BYOL);
        pricingTypes.put(Triple.of(ProServiceProduct, byol, new HashSet<>()),BYOL);
        pricingTypes.put(Triple.of(AiqProduct, byol, new HashSet<>()),BYOL);
        pricingTypes.put(Triple.of(MLProduct, byol, new HashSet<>()),BYOL);
        pricingTypes.put(Triple.of(ContainerProduct, byol, new HashSet<>()),BYOL);
        pricingTypes.put(Triple.of(DataProduct, byol, new HashSet<>()),BYOL);
        pricingTypes.put(Triple.of(ContainerProduct, legal, new HashSet<>()),Free);
        pricingTypes.put(Triple.of(AmiProduct, freeTrialUsage, new HashSet<>()),FTH);
        pricingTypes.put(Triple.of(ContainerProduct, freeTrialUsage, new HashSet<>()),FTH);
        pricingTypes.put(Triple.of(MLProduct, freeTrialUsage, new HashSet<>()),FTH);

        Triple<String, Set<String>, Set<String>> key = Triple.of(productType, agreementTermType, offerType);
        
        if (pricingTypes.containsKey(key)) {
            return pricingTypes.get(key);
        } else {
            return "Unknown";
        }
    }
    
    public static String needToCheckOfferTermsType(String productType, Set<String>agreementTermTypes) {
    	Map<KeyPair, String> offerTermTypes = new HashMap<>();
    	offerTermTypes.put(new KeyPair(ContainerProduct, config), "Y");
    	offerTermTypes.put(new KeyPair(AmiProduct, config), "Y");
    	offerTermTypes.put(new KeyPair(ContainerProduct, fixedPayment), "Y");
    	offerTermTypes.put(new KeyPair(AmiProduct, fixedPayment), "Y");

    	KeyPair key = new KeyPair(productType, agreementTermTypes);
    	if (offerTermTypes.containsKey(key)) {
            return offerTermTypes.get(key);
        } else {
            return null; // or some other appropriate value
        }
    }
    
    public static List<AgreementViewSummary> getAgreementsById () {
    	List<AgreementViewSummary> agreementSummaryList = new ArrayList<AgreementViewSummary>();
		
		Filter partyType = Filter.builder()
                .name("PartyType")
                .values("Proposer")
                .build();
		
		Filter agreementType = Filter.builder()
                .name("AgreementType")
                .values("PurchaseAgreement")
                .build();
		
		Filter idFilter = Filter.builder()
                .name(idType)
                .values(idValue)
                .build();
		
		SearchAgreementsRequest searchAgreementsRequest = SearchAgreementsRequest.builder()
                .catalog("AWSMarketplace")
                .filters(partyType, agreementType, idFilter)
                .build();
		
		SearchAgreementsResponse searchResult = mpasClient.searchAgreements(searchAgreementsRequest);
		
		agreementSummaryList.addAll(searchResult.agreementViewSummaries());
		
		while (searchResult.nextToken() != null && searchResult.nextToken().length() > 0 ) {
			searchAgreementsRequest = SearchAgreementsRequest.builder()
	                .catalog("AWSMarketplace")
	                .filters(partyType, agreementType)
	                .nextToken(searchResult.nextToken())
	                .build();
			searchResult = mpasClient.searchAgreements(searchAgreementsRequest);
			agreementSummaryList.addAll(searchResult.agreementViewSummaries());
		}
		return agreementSummaryList;
    	
    }
    
    static class KeyPair {
        private final String first;
        private final Set<String> second;

        public KeyPair(String productType, Set<String> second) {
            this.first = productType;
            this.second = second;
        }

        @Override
        public int hashCode() {
            // Implement a suitable hashCode() method
            // Combine hash codes of 'first' and 'second'
            return Objects.hash(first, second);
        }

        @Override
        public boolean equals(Object obj) {
            // Implement a suitable equals() method
            // Compare 'first' and 'second' fields for equality
            if (this == obj) return true;
            if (obj == null || getClass() != obj.getClass()) return false;
            KeyPair other = (KeyPair) obj;
            return Objects.equals(first, other.first) && Objects.equals(second, other.second);
        }
    }
    
    public static Set<String> getOfferTermTypes(String offerId ) {
    	
    	Set<String> offerTermTypes = new HashSet<String>();
		
    	DescribeEntityRequest request = DescribeEntityRequest.builder().catalog("AWSMarketplace").entityId(offerId).build();

		DescribeEntityResponse result = awsMarketplaceCatalog.describeEntity(request);

        String details = result.details();
        
        try {
            ObjectMapper objectMapper = new ObjectMapper();
            JsonNode rootNode = objectMapper.readTree(details);
            JsonNode termsNode = rootNode.get("Terms");

            for (JsonNode termNode : termsNode) {
                String type = termNode.get("Type").asText();
                offerTermTypes.add(type);
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
        
		return offerTermTypes;
    	
    }

    public static Set<String> getAgreementTermTypes(GetAgreementTermsResponse agreementTerm) {
    	Set<String> agreementTermTypes = new HashSet<String>();
    	try {
    		for ( AcceptedTerm term : agreementTerm.acceptedTerms()) {
    			ObjectMapper objectMapper = new ObjectMapper();
    	        JsonNode termNode = objectMapper.readTree(getJson(term));
    	        Iterator<Map.Entry<String, JsonNode>> fieldsIterator = termNode.fields();
                while (fieldsIterator.hasNext()) {
                    Map.Entry<String, JsonNode> entry = fieldsIterator.next();
                    JsonNode value = entry.getValue();
                    if (value.isObject() && value.has("type")) {
                    	System.out.println("type = " + value.get("type"));
                    	agreementTermTypes.add(value.get("type").asText());
                    }
                }
    		}
        } catch (Exception e) {
            e.printStackTrace();
        }
		return agreementTermTypes;
    	
    }
    
    //make sure all elements in array2 exist in array1
    public static boolean allElementsExist(Set<String> array1, Set<String> array2) {
        for (String element : array2) {
            boolean found = false;
            for (String str : array1) {
                if (element.equals(str)) {
                    found = true;
                    break;
                }
            }
            if (!found) {
                return false;
            }
        }
        return true;
    }
    
    public static Set<String> getMatchedTermTypesCombination(Set<String> agreementTermTypes) {
    	Set<String> matchedCombination = new HashSet<String>();
    	for ( Set<String> element : allAgreementTypesCombination) {
    		if ( allElementsExist(agreementTermTypes,element)) {
    			matchedCombination = element;
    		}
    	}
    	return matchedCombination;
    }
    
    
	public static void main(String[] args) {

		List<AgreementViewSummary> agreements = getAgreementsById();
		
		for (AgreementViewSummary summary : agreements) {
			String pricingType = "";
			String agreementId = summary.agreementId();
			System.out.println(agreementId);
			String offerId = summary.proposalSummary().offerId();
			Set<String> offerTermTypes = getOfferTermTypes(offerId);
			String prodType = summary.proposalSummary().resources().get(0).type();
			GetAgreementTermsRequest agreementTermsRequest = GetAgreementTermsRequest.builder()
					.agreementId(agreementId)
					.build();
			GetAgreementTermsResponse result = mpasClient.getAgreementTerms(agreementTermsRequest);
			Set<String> agreementTermTypes = getAgreementTermTypes(result);
			Set<String> matchedTermType = getMatchedTermTypesCombination(agreementTermTypes);
			String needToCheckOfferType = needToCheckOfferTermsType(prodType,matchedTermType);
			if ( needToCheckOfferType != null ) {
				Set<String> offerMatchedTermType = getMatchedTermTypesCombination(offerTermTypes);
				pricingType = getPricingType(prodType, matchedTermType, offerMatchedTermType);
			} else if (matchedTermType == legal ) {
				pricingType = Free;
			} else {
				pricingType = getPricingType(prodType, matchedTermType, new HashSet());
			}
			System.out.println("Pricing type is = " + pricingType);
		}
	}

	private static String getJson(Object result) {
		String json = "";
			
		try {
			ObjectMapper om = new ObjectMapper();
			om.setVisibility(PropertyAccessor.FIELD, Visibility.ANY);
			om.registerModule(new JavaTimeModule());
			ObjectWriter ow = om.writer().withDefaultPrettyPrinter();
			
			json = ow.writeValueAsString(result);
		} catch (JsonProcessingException e) {
			e.printStackTrace();
		}
		return json;
	}
	

}
